### PR TITLE
Moving/Updating the WasmPlugin tutorial and Adding ImagePullPolicy description

### PR DESCRIPTION
**Pull request by ingwonsong**
This PR updates the existing WasmPlugin tutorial contents as follows:

* Move the existing doc to `tasks` from `ops`: The guidance is more related to a specific `task` rather than generic operations.
* Add tests for the doc
* Re-org the structure in the doc

In addition, the explanation for the `ImagePullPolicy` is added under `ops/configuration/extensibility`

### CODE REVIEW

File: content/en/docs/tasks/extensibility/wasm-module-distribution/index.md
Comment (craigbox) [inline content/en/docs/tasks/extensibility/wasm-module-distribution/index.md:77]:
```suggestion
## Clean up Wasm modules
```

**Comment by ingwonsong** [comment]
cc: @mathetake @dgn

File: content/en/docs/tasks/extensibility/_index.md
Comment (craigbox) [inline content/en/docs/tasks/extensibility/_index.md:3]:
```suggestion
description: Demonstrates how to distribute and configure a Wasm module.
```

File: content/en/docs/tasks/extensibility/_index.md
Comment (douglas-reid) [inline content/en/docs/tasks/extensibility/_index.md:8]: if we add this, do we need the other links changes in the blog posts, etc. ?

File: content/en/docs/tasks/extensibility/_index.md
Comment (douglas-reid) [inline content/en/docs/tasks/extensibility/_index.md:3]: wdyt about leaving the possibility open for more than Wasm modules here?

Maybe: "Demonstrates how to extend mesh behavior." ?

**Review by douglas-reid — COMMENTED** [review]
Thanks for tackling this.

File: content/en/docs/tasks/extensibility/_index.md
Comment (ingwonsong) [reply content/en/docs/tasks/extensibility/_index.md:3]: Seems to make sense. Will change this.

Comment (ingwonsong) [reply content/en/docs/tasks/extensibility/_index.md]: Done.

**Comment by ingwonsong** [comment]
@craigbox @douglas-reid I added new doc `content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md`. Please take a look into it as well. Thanks!

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:25]:
```suggestion

When `ImagePullPolicy` is not specified for a resource, Istio defaults to `IfNotPresent` behavior. However, if the provided `url` field specifies an OCI image that has a tag value of `latest`, Istio will use `Always` behavior.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:21]:
```suggestion
Users explicitly configure the behavior for Wasm module retrieval with the `ImagePullPolicy` field. However, user-provided behavior can be overridden by Istio in the following scenarios:
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:36]:
```suggestion
In Kubernetes, `ImagePullPolicy: Always` means that an image is pulled directly from its source each time a pod is created.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:37]:
```suggestion
Every time of a new pod is spanned, Kubernetes pulls the image anew.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:39]:
```suggestion
In case of `WasmPlugin`, `ImagePullPolicy: Always` means that Istio will pull an image directly from its source each time the corresponding `WasmPlugin` Kubernetes resource is created or changed.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:40]:
```suggestion
Please note that a change not only in `spec` but also `metadata` triggers pulling of Wasm module when `Always` policy is used. This can mean that an image is pulled from source several times over the lifetime of a pod and over the lifetime of an individual proxy.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (douglas-reid) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:24]:
```suggestion
1. If the `url` field points to an OCI image and it has a digest suffix (e.g., `gcr.io/foo/bar@sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef`), `IfNotPresent` policy is used.
```

**Comment by zirain** [comment]
what does this mean?
image

**Review by zirain — APPROVED** [review]
lgtm, would be better add more contents about `VmConfig` which I think it's useful to end-users.

**Comment by ingwonsong** [comment]
As you know, `VmConfig` is basically a part of Envoy API.
So, IMHO, the ideal place of such contents would be Envoy doc.

WDYT?

**Comment by ingwonsong** [comment]
Yes. sounds good to me.

**Comment by ingwonsong** [comment]
@craigbox If there is any further needed LGTM, could you please guide me?

**Comment by ingwonsong** [comment]
> what does this mean? image

This seems to mean "Alpha" status. If we set `status: Alpha` in the header of each doc, the star is added.

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (craigbox) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:3]:
```suggestion
description: Describes how Istio determines whether to pull Wasm modules or use cached versions.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (craigbox) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:7]:
```suggestion
```
never had another location so can remove these

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (craigbox) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:15]:
```suggestion
The [WasmPlugin API](/docs/reference/config/proxy_extensions/wasm-plugin) provides a method for [distributing Wasm modules](/docs/tasks/extensibility/wasm-module-distribution) to proxies.
Since each proxy will pull Wasm modules from a remote registry or an HTTP server, understanding how Istio chooses to pull modules is important in terms of usability as well as performance.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (craigbox) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:33]:
```suggestion
## Lifecycle of cached modules

Each proxy, whether a sidecar proxy or a gateway, caches Wasm modules. The lifetime of the cached Wasm module is therefore bounded by the lifetime of the corresponding pod.
In addition, there is an expiration mechanism for keeping the proxy memory footprint to a minimum: if a cached Wasm module is not used for a certain amount of the time, the module is purged.

This expiration can be configured via the environment variables `WASM_MODULE_EXPIRY` and `WASM_PURGE_INTERVAL` of [pilot-proxy](/docs/reference/commands/pilot-agent/#envvars), which are the duration of expiration and the interval for checking the expiration respectively.
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (craigbox) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:41]:
```suggestion
## The meaning of "Always"

In Kubernetes, `ImagePullPolicy: Always` means that an image is pulled directly from its source each time a pod is created.
Every time a new pod is started, Kubernetes pulls the image anew.

For a `WasmPlugin`, `ImagePullPolicy: Always` means that Istio will pull an image directly from its source each time the corresponding `WasmPlugin` Kubernetes resource is created or changed.
Please note that a change not only in `spec` but also `metadata` triggers the pulling of a Wasm module when the `Always` policy is used. This can mean that an image is pulled from source several times over the lifetime of a pod, and over the lifetime of an individual proxy.
```

File: content/en/docs/tasks/extensibility/wasm-module-distribution/index.md
Comment (craigbox) [inline content/en/docs/tasks/extensibility/wasm-module-distribution/index.md:31]: Is this format required for doc tests?

File: content/en/docs/tasks/extensibility/wasm-module-distribution/index.md
Comment (craigbox) [inline content/en/docs/tasks/extensibility/wasm-module-distribution/index.md:58]:
```suggestion
{{< idea >}}
If a `WasmPlugin` is created in a specific namespace besides `istio-system`, the pods in that namespace will be configured. If the resource is created in the `istio-system` namespace, all namespaces will be affected.
{{< /idea >}}
```

File: content/en/docs/tasks/extensibility/wasm-module-distribution/index.md
Comment (craigbox) [inline content/en/docs/tasks/extensibility/wasm-module-distribution/index.md:62]:
```suggestion
1. Test `/productpage` without credentials
```

File: content/en/docs/tasks/extensibility/wasm-module-distribution/index.md
Comment (craigbox) [inline content/en/docs/tasks/extensibility/wasm-module-distribution/index.md:69]:
```suggestion
1. Test `/productpage` with credentials
```

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:19]: I'd opt for `Inspired by` or `Analogous to`

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:38]: `spawned`

Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md:38]: also maybe something like

```suggestion
Every time of a new pod is spanned, Kubernetes pulls the image anew. As istio does not use your cluster's container runtime to pull the image, subsequent pulls on the same node will not be faster than the first. Istio will always pull the full image, no matter if it has pulled an image with the same SHA previously. Keep this in mind when using `ImagePullPolicy: Always`. 
```

File: content/en/docs/tasks/extensibility/wasm-module-distribution/index.md
Comment (ingwonsong) [reply content/en/docs/tasks/extensibility/wasm-module-distribution/index.md:31]: Yes.

File: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md
Comment (ingwonsong) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md]:
```suggestion
Every time of a new pod is spawned, Kubernetes pulls the image anew.
```

Comment (ingwonsong) [inline content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md]:
```suggestion
Every time of a new pod is spawned, Kubernetes pulls the image anew.
```

Comment (ingwonsong) [reply content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md]: Oh... I understood.

Thing is, in current implementation, we check the manifest first, and then determine whether to pull or just to use the cached Wasm binary. So, the difference with K8S is only that they are cached in "proxy instance" not "node".

Since this is explained in the previous section, IMHO, it seems to be not necessary to add this part.

WDYT?

Comment (ingwonsong) [reply content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md]: At this time, let's go with the original sentence.

@dgn Anytime, please let me know if you have any concern here. Thanks!

**Review by craigbox — APPROVED** [review]
LGTM, approving and setting `do-not-merge` in case anything else needs to be updated before merging.

/cherrypick release-1.14

**Pull request merged**
